src/stan/variational/advi.hpp

```diff
@@ -0,0 +1,665 @@
+          for (int d = 0; d < dim; ++d) {
+            z_check(d) = stan::prob::normal_rng(0, 1, rng_);
+          }
+          z_tilde = advi_params.to_unconstrained(z_check);
+
+          // Accumulate log probability
+          elbo += (model_.template
+                   log_prob<false, true>(z_tilde, print_stream_));
+        }
+        // Divide to get Monte Carlo integral estimate
+        elbo /= static_cast<double>(n_monte_carlo_elbo_);
+
+        // Add entropy term
+        elbo += advi_params.entropy();
+
+        return elbo;
+      }
+
+      /**
+       * FULL-RANK GRADIENTS
+       *
+       * Calculates the "blackbox" gradient with respect to BOTH the location
+       * vector (mu) and the cholesky factor of the scale matrix (L) in
+       * parallel. It uses the same gradient computed from a set of Monte Carlo
+       * samples
+       *
+       * @param muL     mean and cholesky factor of affine transform
+       * @param mu_grad gradient of location vector parameter
+       * @param L_grad  gradient of scale matrix parameter
+       */
+      void calc_combined_grad(const advi_params_fullrank& muL,
+                              Eigen::VectorXd& mu_grad,
+                              Eigen::MatrixXd& L_grad) {
+        static const char* function =
+          "stan::variational::advi.calc_combined_grad";
+
+        int dim       = muL.dimension();
+        double tmp_lp = 0.0;
+
+        stan::math::check_size_match(function,
+                        "Dimension of muL", muL.dimension(),
+                        "Dimension of variables in model", cont_params_.size());
+        stan::math::check_size_match(function,
+                        "Dimension of mu grad vector", mu_grad.size(),
+                        "Dimension of mean vector in variational q", dim);
+        stan::math::check_square(function, "Scale matrix", L_grad);
+        stan::math::check_size_match(function,
+                        "Dimension of scale matrix", L_grad.rows(),
+                        "Dimension of mean vector in variational q", dim);
+
+        // Initialize everything to zero
+        mu_grad = Eigen::VectorXd::Zero(dim);
+        L_grad  = Eigen::MatrixXd::Zero(dim, dim);
+        Eigen::VectorXd tmp_mu_grad = Eigen::VectorXd::Zero(dim);
+        Eigen::VectorXd z_check = Eigen::VectorXd::Zero(dim);
+        Eigen::VectorXd z_tilde = Eigen::VectorXd::Zero(dim);
+
+        // Naive Monte Carlo integration
+        for (int i = 0; i < n_monte_carlo_grad_; ++i) {
+          // Draw from standard normal and transform to unconstrained space
+          for (int d = 0; d < dim; ++d) {
+            z_check(d) = stan::prob::normal_rng(0, 1, rng_);
+          }
+          z_tilde = muL.to_unconstrained(z_check);
+
+          // Compute gradient step in unconstrained space
+          stan::model::gradient(model_, z_tilde, tmp_lp, tmp_mu_grad,
+                                print_stream_);
+          stan::agrad::recover_memory();
```

Comment:
> On May 1, 2015, at 8:41 AM, Alp Kucukelbir <notifications@github.com> wrote:
> 
> In src/stan/variational/advi.hpp:
> 
> > +        Eigen::VectorXd tmp_mu_grad = Eigen::VectorXd::Zero(dim);
> > +        Eigen::VectorXd z_check = Eigen::VectorXd::Zero(dim);
> > +        Eigen::VectorXd z_tilde = Eigen::VectorXd::Zero(dim);
> > +
> > +        // Naive Monte Carlo integration
> > +        for (int i = 0; i < n_monte_carlo_grad_; ++i) {
> > +          // Draw from standard normal and transform to unconstrained space
> > +          for (int d = 0; d < dim; ++d) {
> > +            z_check(d) = stan::prob::normal_rng(0, 1, rng_);
> > +          }
> > +          z_tilde = muL.to_unconstrained(z_check);
> > +
> > +          // Compute gradient step in unconstrained space
> > +          stan::model::gradient(model_, z_tilde, tmp_lp, tmp_mu_grad,
> > +                                print_stream_);
> > +          stan::agrad::recover_memory();
> 
> i put this in after a bug-hunt with daniel. without it, there is a memory leak that grows across iterations. hope this helps?

Could you check that it still leaks without recover_memory()?  If so,
it's a deeper bug.  You should *not* need to call recover_memory() after
calling this gradient() function.

- Bob

P.S.  stan/model/util.hpp   for stan::model::gradient wrapper of functional

      stan/math/rev/mat/functor/gradient.hpp   for functional definition




> and thanks for all the cleanup work!
> …
> On Apr 30, 2015, at 18:34, Rob Trangucci <notifications@github.com> wrote: In src/stan/variational/advi.hpp: > + Eigen::VectorXd tmp_mu_grad = Eigen::VectorXd::Zero(dim); > + Eigen::VectorXd z_check = Eigen::VectorXd::Zero(dim); > + Eigen::VectorXd z_tilde = Eigen::VectorXd::Zero(dim); > + > + // Naive Monte Carlo integration > + for (int i = 0; i < n_monte_carlo_grad_; ++i) { > + // Draw from standard normal and transform to unconstrained space > + for (int d = 0; d < dim; ++d) { > + z_check(d) = stan::prob::normal_rng(0, 1, rng_); > + } > + z_tilde = muL.to_unconstrained(z_check); > + > + // Compute gradient step in unconstrained space > + stan::model::gradient(model_, z_tilde, tmp_lp, tmp_mu_grad, > + print_stream_); > + stan::agrad::recover_memory(); @bob-carpenter do we need this? model::gradient calls agrad::gradient which calls start_memory_nested() and recover_memory_nested(). — Reply to this email directly or view it on GitHub.
> —
> Reply to this email directly or view it on GitHub.
>